src/bootstrap_cache.rs

```diff
@@ -0,0 +1,163 @@
+use crate::dirs::Dirs;
+use crate::{Error, NodeInfo, R};
+use bincode::{deserialize_from, serialize_into};
+use log::{error, info};
+use std::fs::{self, File};
+use std::io::{self, BufReader, BufWriter};
+use std::path::Path;
+use std::path::PathBuf;
+
+/// Maximum peers in the cache.
+const MAX_CACHE_SIZE: usize = 200;
+
+type Stack = Vec<NodeInfo>;
+
+/// A very simple LRU like struct that writes itself to disk every 10 entries added.
+pub struct BootstrapCache {
+    stack: Stack,
+    path_buf: PathBuf,
+    add_count: u8,
+}
+
+impl BootstrapCache {
+    pub fn try_new(dirs: &Dirs) -> R<BootstrapCache> {
+        let cache_file = path(dirs);
+        let stack = if cache_file.exists() {
+            read_from_disk(&cache_file)?
+        } else {
+            let cache_dir = cache_file
+                .parent()
+                .ok_or_else(|| io::ErrorKind::NotFound.into())
+                .map_err(Error::IoError)?;
+            fs::create_dir_all(&cache_dir)?;
+            Default::default()
+        };
+
+        Ok(BootstrapCache {
+            stack,
+            path_buf: cache_file,
+            add_count: 0u8,
+        })
+    }
+
+    #[allow(unused)]
+    pub fn peers(&self) -> &Stack {
+        &self.stack
+    }
+
+    pub fn add_peer(&mut self, peer: NodeInfo) {
+        if self.stack.contains(&peer) {
+            self.move_to_cache_top(peer);
+        } else {
+            self.insert_new(peer);
+        }
+    }
+
+    fn insert_new(&mut self, peer: NodeInfo) {
+        self.stack.push(peer);
+        self.add_count += 1;
+        if self.stack.len() == MAX_CACHE_SIZE {
+            let _ = self.stack.pop();
+        }
+        self.try_sync_to_disk();
+    }
+
+    fn move_to_cache_top(&mut self, peer: NodeInfo) {
+        // remove and push back
+        let mut t: Stack = self.stack.iter().filter(|&x| *x == peer).cloned().collect();
+        t.push(peer);
+        self.stack = t;
```

Comment:
Seems a very wasteful way to achieve the goal of just putting someone to the top. Why not just find and remove from that index and push it to the same vec instead of a fresh new allocation etc.

So
```
if let Some(info) = stack.remove(n) { stack.push(info); }
// n can be obtained via `pos()` or check if you can find and remove at the same time
```

Goes without saying if `n == len - 1` then you don't really have to do anything.